Simplify guest linker configuration

shkoo opened the pull request:
* Remove custom linker script in favor of -Ttext= linker arg
* Stack pointer and .text are now based on values in memory.rs instead of a separate linker script.  .data and .bss follow .text.
* When using "std" on the guest, we no longer need to include #![no_main] or risc0_zkvm::entry!(...); just a standard "pub fn main" will do. (Unfortunately we still have to force inclusion of the risc0 zkvm crate with "use risc0_zkvm as _" if we don't use anything else from it)

shkoo commented:
> When building full projects, would users now need to use `cargo risczero build/test` instead of `cargo build/test`? Or are these just utilities for building/testing pure guest code (i.e. in the absence of host code)?

These are just utilities for building/testing pure guest code in the absence of host code.

Going this direction could support the other use case too though, if that's the direction we want to go!

tzerrell commented:
> These are just utilities for building/testing pure guest code in the absence of host code.
> 
> Going this direction could support the other use case too though, if that's the direction we want to go!

Great, that seems like a useful feature to have available! IMO, in the absence of technical benefits/problems to the two approaches, using `cargo build` for a "full" project is better than `cargo risczero build` -- but if we gain technical benefits to the latter it could be worthwhile. That's beside the point for this PR regardless though.

flaub commented on risc0/zkvm/src/guest/mod.rs:
```diff
@@ -154,27 +155,32 @@ macro_rules! entry {
-    }
+    env::init();
 
-    let mut bss = __bss_begin;
-    while bss < __bss_end {
-        ptr::write_volatile(bss, mem::zeroed());
-        bss = bss.offset(1);
```
I think we need to be able to prove that the guest initialized the BSS

shkoo commented:
After discussion with Frank and Jeremy, I'm going to refactor and work on doing the following:

* Put TEXT at 0x400, stack at top of address space, everything else packed after TEXT.
* Check with @tzerrell to make sure the no_main changes are OK

tzerrell commented:
>     * Check with @tzerrell to make sure the no_main changes are OK

The `no_main` changes in the examples in the current version of this PR seemed reasonable to me. What I wonder (and probably this is worth a discussion with you) is what other ramifications there are beyond what's apparent in the examples -- in particular if there is any impact on integration with Bonsai, as that is simultaneously the most important and the most different from `examples`.

shkoo commented:
> The impact of the `no_main` changes in the examples in the current version of this PR seemed reasonable to me. What I wonder (and probably this is worth a discussion with you) is what other ramifications there are beyond what's apparent in the examples -- in particular if there is any impact on integration with Bonsai, as that is simultaneously the most important and the most different from `examples`.

I changed the entry code to allow "risc0_zkvm::entry!" even when in "std" mode, and reverted the changes in "examples" since we don't need them anymore, so we should be good on that front.

This should have no effect on any bonsai integration.

This should be ready to review now!

nategraf commented:
On 1 and 2, both were originally included in the draft PR I put up mostly for visibility. Maybe we split that off?

shkoo commented on risc0/build/src/lib.rs:
```diff
@@ -391,6 +392,13 @@ fn build_guest_package<P>(
                 // Remap absolute pathnames in compiled ELFs for builds that are more reproducible.
                 "-Z",
                 "remap-cwd-prefix=.",
+                // Specify where to start loading the program in memory.
+                "-C",
+                &format!("link-arg=-Ttext=0x{:08X}", memory::TEXT_START),
```
They go after the .text section, as we discussed.

flaub replied:
How do we ensure that the data goes to these locations? If we're not specifying it where does this information come from?

shkoo replied:
This is the default behavior of the linker unless it's overridden by e.g. a -Tdata argument or a custom build script.

flaub replied:
Is there a link or reference to describe this behavior that we can add as a comment? I'm worried that this kind of thing is very arcane and if someone needs to go and understand how this all works, we need to have some way to get all the context necessary to support it.

shkoo commented on risc0/zkvm/platform/src/syscall.rs:
```diff
@@ -494,20 +494,43 @@ pub unsafe extern "C" fn sys_getenv(
+        extern "C" {
+            // This symbol marks the end of all elf sections, so this is where we start our
+            // heap.
+            static _end: u32;
```
https://github.com/llvm/llvm-project/blob/76941b68ecd9a1f9ca337774b4082c2ca7577450/lld/ELF/Writer.cpp#L247

Tried to clarify the comment a little bit.

flaub replied:
Is this specific to LLVM or do all C-like compilers do this?

shkoo replied:
Best I can tell, it's standard for loaders to do this, and e.g. newlib depends on it.

flaub replied:
Can you add a link to newlib so that later if we see this code we can get some context about it?

shkoo commented on risc0/build/src/lib.rs:
```diff
@@ -142,19 +141,17 @@ fn sha_digest_with_hex(data: &[u8]) -> (Vec<u8>, String) {
-where
-    P: AsRef<Path>,
-{
+#[doc(hidden)]
```
> 1. We should ban `#[doc(hidden)]` or have clear justification for its use.
> 2. Each PR should try to aim to address a single thing. This seems to have a mix of linker script changes along with `risczero build` support, which doesn't seem directly dependent on one another.

Ok, removed the "cargo risczero" changes from this PR

> 3. Without an explicit linker script, it's unclear where the DATA section is. There's an `_end` symbol that needs more documentation or a reference/link to how this symbol is defined. Do we know if this will work with all compilers?

I couldn't find any references to cite for it, but it works with LLVM, and as far as I know we don't have plans to support any other rust backends.

Newlib references it as following in a non-platform-specific header file, so it looks pretty standard to me:

`extern char _end[];                /* _end is set in the linker command file */
